pkg/controllers/mpi_job_controller.go

```diff
@@ -480,13 +475,36 @@ func (c *MPIJobController) getLauncherJob(mpiJob *kubeflow.MPIJob) (*batchv1.Job
 	return launcher, nil
 }
 
-// getTotalGPUs gets the total number of desired GPUs. Defaults to 1 if not specified.
-func getTotalGPUs(mpiJob *kubeflow.MPIJob) int {
-	totalGPUs := 1
+// allocateGPUs allocates the worker replicas and GPUs per worker.
+func allocateGPUs(mpiJob *kubeflow.MPIJob, gpusPerNode int, done bool) (workerReplicas int, gpusPerWorker int, err error) {
+	workerReplicas = 0
+	gpusPerWorker = 0
+	err = nil
 	if mpiJob.Spec.GPUs != nil {
-		totalGPUs = int(*mpiJob.Spec.GPUs)
+		totalGPUs := int(*mpiJob.Spec.GPUs)
+		if totalGPUs < gpusPerNode {
+			workerReplicas = 1
+			gpusPerWorker = totalGPUs
+		} else if totalGPUs % gpusPerNode == 0 {
+			workerReplicas = totalGPUs / gpusPerNode
+			gpusPerWorker = gpusPerNode
+		} else {
+			err = fmt.Errorf("specified #GPUs is not a multiple of GPUs per node (%d)", gpusPerNode)
```

Comment:
Could you please explain more about the error? What will happen in this situation?

Reply:
In `MPIJob` all the workers are homogeneous.  So,  in this case, I think user can use `replicas: N` and `resources.limits.'nvidia.com/gpu':M`.

ah,  how about printing "please user `replicas` instead of `gpus`"??

Reply:
Got it, the error message is little confusing because I am not familiar with MPIJob now.

Thanks for your explanation :smile:

Reply:
Yeah it's a little confusing. We actually have two ways of specifying GPUs.
* The "simple" version you just specify `gpus`, the controller then divides that number by the max gpus per node to calculate the # replicas. In this case, if you specify `gpus: 12` while you have 8-gpu node, it would cause this error.
* The "custom" version you can specify the # replicas and # gpus / replica. For example, if you have 8-gpu nodes, you can do 4x4 if you want.

I'm planning to add some validation (#3) to enforce the rules on `gpus`.